docs: add lesson 5 (#6)
microsoft/generative-ai-with-javascript: lessons/05-rag/README.md

# Lesson 5: Talk to your data with Retrieval-Augmented Generation (RAG)

In this lesson, you will learn:

- The basics of Retrieval-Augmented Generation (RAG) and how it can be used to enhance the responses of generative AI models.
- How to integrate external data sources into your AI application.
- To leverage your data to improve the relevance and accuracy of the AI-generated content.

> 

Reviewer: ```suggestion
## Getting Started

Follow these steps to get started with this curriculum:

1. **Fork the Repository**: Select the _fork_ button in the upper right-hand corner of the repository or this link:
   [![Fork](https://img.shields.io/badge/Fork-Repository-blue)](https://github.com/microsoft/generative-ai-with-javascript/fork)

2. **Clone the Repository**: Select the _clone_ button in the upper right-hand corner of the repository or this link:
   [![Clone](https://img.shields.io/badge/Clone-Repository-blue)](https://github.com/microsoft/generative-ai-with-javascript.git)

3. **Open in GitHub Codespaces**: Starting this repository with GitHub Codespaces will enable you to run the same code as it uses [GitHub Models](https://docs.github.com/en/github-models).

    Click the button below to start a Codespace:
   [![Open in GitHub Codespaces](https://github.com/codespaces/badge.svg)](https://github.com/codespaces/new?hide_repo_select=true&ref=main&repo=832180352)
```

# Lesson 5: Talk to your data with Retrieval-Augmented Generation (RAG)

In this lesson, you will learn:

- The basics of Retrieval-Augmented Generation (RAG) and how it can be used to enhance the responses of generative AI models.
- How to integrate external data sources into your AI application.
- To leverage your data to improve the relevance and accuracy of the AI-generated content.

> ## Getting Started

Reviewer: ```suggestion

## Getting Started
```

- How to integrate external data sources into your AI application.
- To leverage your data to improve the relevance and accuracy of the AI-generated content.

## Getting Started

Follow these steps to get started with this curriculum:

1. **Fork the Repository**: Select the _fork_ button in the upper right-hand corner of the repository or this link:
   [![Fork](https://img.shields.io/badge/Fork-Repository-blue)](https://github.com/microsoft/generative-ai-with-javascript/fork)

2. **Clone the Repository**: Select the _clone_ button in the upper right-hand corner of the repository or this link:
   [![Clone](https://img.shields.io/badge/Clone-Repository-blue)](https://github.com/microsoft/generative-ai-with-javascript.git)

3. **Open in GitHub Codespaces**: Starting this repository with GitHub Codespaces will enable you to run the same code as it uses [GitHub Models](https://docs.github.com/en/github-models).

    Click the button below to start a Codespace:
   [![Open in GitHub Codespaces](https://github.com/codespaces/badge.svg)](https://github.com/codespaces/new?hide_repo_select=true&ref=main&repo=832180352)

> ## Related resources

Reviewer: ```suggestion

## Related resources
```

The door creaked open, revealing Ada Lovelace, her eyes sparkling with curiosity. "Ah, it's about time you arrived," she said warmly. "We need you to run an errand."

Charles Babbage stepped forward, examining the time beetle in your hand. "This device is remarkable, but it's a bit faulty, isn't it? You've noticed, I'm sure."

<div>
  <img src="./assets/ada.jpeg" alt="Ada Lovelace and Charles Babbage working on a device" width="400">
</div>

Leonardo nodded, "Sì, it has been acting strangely."

**Ada:** the device isn't quite ready, we need to give it more capabilities. We need to make it smarter, more aware of the world around it. The idea is for it to be able to retrieve information from different time periods and use it to generate responses that are accurate and relevant. Can you help with that?

**You:** of course, sounds like we need to _augment_ the responses of the device with data, makes sense.

**Ada:** Let's talk about a concept I'd like to call RAG, or Retrieval-Augmented Generation.

## Talk to Ada Lovelace

<div>
  <img src="./assets/ada-2.jpeg" alt="Ada Lovelace" width="400">
</div>

If you want to talk to Ada, run the app [Ada Lovelace](./characters/).

> > NOTE: Remember to start a Codespaces and then run `npm install` followed by `npm start`. 

Reviewer: ```suggestion
> NOTE: Start a [![GitHub Codespace](https://img.shields.io/badge/GitHub-Codespace-brightgreen)](https://codespaces.new/microsoft/generative-ai-with-javascript), navigate to _lessons/05-rag/characters_, and then run `npm install` followed by `npm start`. Once it appears, select the "Open in Browser" button.
```
